binding.cc

```diff
@@ -919,1155 +800,1353 @@ sn_crypto_stream_xor(
-  SN_ARGV(4, crypto_kdf_derive_from_key)
-
-  SN_ARGV_TYPEDARRAY(subkey, 0)
-  SN_ARGV_UINT64(subkey_id, 1)
-  SN_ARGV_TYPEDARRAY(ctx, 2)
-  SN_ARGV_TYPEDARRAY(key, 3)
-
-  SN_ASSERT_MIN_LENGTH(subkey_size, crypto_kdf_BYTES_MIN, "subkey")
-  SN_ASSERT_MAX_LENGTH(subkey_size, crypto_kdf_BYTES_MAX, "subkey")
-  SN_ASSERT_LENGTH(ctx_size, crypto_kdf_CONTEXTBYTES, "ctx")
-  SN_ASSERT_LENGTH(key_size, crypto_kdf_KEYBYTES, "key")
-
-  SN_RETURN(crypto_kdf_derive_from_key(subkey_data, subkey_size, subkey_id, (const char *) ctx_data, key_data), "could not generate key")
+static inline int
+sn_crypto_kdf_derive_from_key(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> subkey,
+  int64_t subkey_id,
+  js_typedarray_span_t<> ctx,
+  js_typedarray_span_t<> key
+) {
+  assert(subkey.size_bytes() >= crypto_kdf_BYTES_MIN);
+  assert(subkey.size_bytes() <= crypto_kdf_BYTES_MAX);
+  assert(ctx.size_bytes() == crypto_kdf_CONTEXTBYTES);
+  assert(key.size_bytes() == crypto_kdf_KEYBYTES);
+
+  return crypto_kdf_derive_from_key(
+    subkey.data(),
+    subkey.size_bytes(),
+    static_cast<uint64_t>(subkey_id),
+    reinterpret_cast<const char *>(ctx.data()),
+    key.data()
+  );
 }
 
-js_value_t *
-sn_crypto_hash (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(2, crypto_hash_sha256)
-
-  SN_ARGV_TYPEDARRAY(out, 0)
-  SN_ARGV_TYPEDARRAY(in, 1)
-
-  SN_ASSERT_LENGTH(out_size, crypto_hash_BYTES, "out")
+static inline int
+sn_crypto_hash(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> out,
+  js_typedarray_span_t<> in
+) {
+  assert(out.size_bytes() == crypto_hash_BYTES);
 
-  SN_RETURN(crypto_hash(out_data, in_data, in_size), "could not compute hash")
+  return crypto_hash(out.data(), in.data(), in.size_bytes());
 }
 
-js_value_t *
-sn_crypto_hash_sha256 (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(2, crypto_hash_sha256)
-
-  SN_ARGV_TYPEDARRAY(out, 0)
-  SN_ARGV_TYPEDARRAY(in, 1)
-
-  SN_ASSERT_LENGTH(out_size, crypto_hash_sha256_BYTES, "out")
+static inline int
+sn_crypto_hash_sha256(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> out,
+  js_typedarray_span_t<> in
+) {
+  assert(out.size_bytes() == crypto_hash_sha256_BYTES);
 
-  SN_RETURN(crypto_hash_sha256(out_data, in_data, in_size), "could not compute hash")
+  return crypto_hash_sha256(out.data(), in.data(), in.size_bytes());
 }
 
-js_value_t *
-sn_crypto_hash_sha256_init (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(1, crypto_hash_sha256_init)
-
-  SN_ARGV_BUFFER_CAST(crypto_hash_sha256_state *, state, 0)
-
-  SN_THROWS(state_size != sizeof(crypto_hash_sha256_state), "state must be 'crypto_hash_sha256_STATEBYTES' bytes")
+static inline int
+sn_crypto_hash_sha256_init(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> state
+) {
+  assert(state.size_bytes() == sizeof(crypto_hash_sha256_state));
+  auto state_data = reinterpret_cast<crypto_hash_sha256_state *>(state.data());
```

Comment:
This could be improved by `libjstl` adding `js_typedarray_span_of_t<T>` similar to `js_arraybuffer_span_of_t<T>`.